File: Suspension matlab/Damping/Damped_oscillation_analysis/Damped_oscillation_analysis.m
%% Damping Ratio Sweep
% Displacement + Tire Normal Load
clear; clc; close all;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end

%% Parameters

ms = damped_ms;
mu = damped_mu;

ks = damped_ks;
kt = damped_kt;

g = damped_g;

%% Damping ratios

zeta_list = damped_zeta_list;

%% Critical damping

cc = 2*sqrt(ks*ms);

%% Initial conditions

xs0 = damped_xs0;
vs0 = damped_vs0;

xu0 = damped_xu0;
vu0 = damped_vu0;

X0 = [xs0; vs0; xu0; vu0];

%% Time

tspan = damped_tspan;

%% Static tire load

Fz_static = (ms + mu)*g;

%% =========================================
% Plot
%% =========================================

figure;

%% =========================================
% Sprung displacement
%% =========================================

subplot(2,1,1)
hold on;
grid on;

%% =========================================
% Tire load variation
%% =========================================

subplot(2,1,2)
hold on;
grid on;

%% =========================================
% Loop
%% =========================================

for i = 1:length(zeta_list)

    zeta = zeta_list(i);

    cs = zeta * cc;

    %% Solve

    [t,X] = ode45( ...
        @(t,X) quarter_car_damped_ode( ...
        t,X,ms,mu,ks,kt,cs), ...
        tspan,X0);

    %% Extract states

    xs = X(:,1);
    xu = X(:,3);

    %% Road profile
    % fixed ground for free vibration

    xr = zeros(size(t));

    %% Tire normal load

    Fz = Fz_static + kt .* (xu - xr);

    %% Dynamic tire load variation

    dFz = Fz - Fz_static;

    %% =====================================
    % Plot displacement
    %% =====================================

    subplot(2,1,1)

    plot(t,xs,'LineWidth',2);

    %% =====================================
    % Plot tire load variation
    %% =====================================

    subplot(2,1,2)

    plot(t,dFz,'LineWidth',2);

end

%% =========================================
% Labels
%% =========================================

subplot(2,1,1)

xlabel('Time (s)');
ylabel('Sprung Displacement (m)');

title('Damping Ratio Sweep');

legend( ...
    '\zeta = 0', ...
    '\zeta = 0.2', ...
    '\zeta = 0.4', ...
    '\zeta = 0.7', ...
    '\zeta = 1.0');

%% =========================================

subplot(2,1,2)

xlabel('Time (s)');
ylabel('\Delta Fz (N)');

title('Dynamic Tire Load Variation');

legend( ...
    '\zeta = 0', ...
    '\zeta = 0.2', ...
    '\zeta = 0.4', ...
    '\zeta = 0.7', ...
    '\zeta = 1.0');

%% =========================================
% ODE Function
%% =========================================

function dX = quarter_car_damped_ode( ...
    ~,X,ms,mu,ks,kt,cs)

%% States

xs = X(1);
vs = X(2);

xu = X(3);
vu = X(4);

%% Relative displacement / velocity

dx = xs - xu;
dv = vs - vu;

%% Accelerations

as = -(ks/ms)*dx ...
     -(cs/ms)*dv;

au =  (ks/mu)*dx ...
     +(cs/mu)*dv ...
     -(kt/mu)*xu;

%% State derivatives

dX = zeros(4,1);

dX(1) = vs;
dX(2) = as;

dX(3) = vu;
dX(4) = au;

end
 

File: Suspension matlab/Damping/Damped_oscillation_analysis/Damped_oscillation_analysis.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np
from scipy.integrate import solve_ivp

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def quarter_car_damped_ode(_t, x, ms, mu, ks, kt, cs):
    xs, vs, xu, vu = x
    dx = xs - xu
    dv = vs - vu
    a_s = -(ks / ms) * dx - (cs / ms) * dv
    a_u = (ks / mu) * dx + (cs / mu) * dv - (kt / mu) * xu
    return [vs, a_s, vu, a_u]


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    ms = p["damped_ms"]
    mu = p["damped_mu"]
    ks = p["damped_ks"]
    kt = p["damped_kt"]
    g = p["damped_g"]
    zeta_list = np.asarray(p["damped_zeta_list"], dtype=float)
    x0 = [p["damped_xs0"], p["damped_vs0"], p["damped_xu0"], p["damped_vu0"]]
    tspan = np.asarray(p["damped_tspan"], dtype=float)
    t_eval = np.linspace(tspan[0], tspan[-1], 1000)

    cc = 2 * np.sqrt(ks * ms)
    fz_static = (ms + mu) * g
    fig, axes = plt.subplots(2, 1)

    for zeta in zeta_list:
        cs = zeta * cc
        sol = solve_ivp(
            quarter_car_damped_ode,
            (tspan[0], tspan[-1]),
            x0,
            t_eval=t_eval,
            args=(ms, mu, ks, kt, cs),
        )
        t = sol.t
        xs = sol.y[0]
        xu = sol.y[2]
        dfz = fz_static + kt * xu - fz_static
        label = f"zeta = {zeta:g}"
        axes[0].plot(t, xs, linewidth=2, label=label)
        axes[1].plot(t, dfz, linewidth=2, label=label)

    axes[0].grid(True)
    axes[0].set_xlabel("Time (s)")
    axes[0].set_ylabel("Sprung Displacement (m)")
    axes[0].set_title("Damping Ratio Sweep")
    axes[0].legend()
    axes[1].grid(True)
    axes[1].set_xlabel("Time (s)")
    axes[1].set_ylabel("Delta Fz (N)")
    axes[1].set_title("Dynamic Tire Load Variation")
    axes[1].legend()
    show_plots()


if __name__ == "__main__":
    main()
